### ISSUE TITLE
Shadows and Highlights

### ISSUE THREAD

**Issue by masc4ii** [ilia3101/MLV-App#71]
@ilia3101 : lets discuss what you need for it here - and lets also discuss when it is ready :-D

I think we need 2 additional sliders? Highlights and Shadows? Or 4? Black and White?

**Comment by bouncyball-git** [comment]
Current sliders are quite clever and can affect almost whole range of the histogram. But they have their quirks, sometimes they are very rough and not precise. If tone mapping is not used pulling down highlights a bit tricky, needs pulling back quite amount of exposure and then correcting whole range by Dark/Light Strength/Range and compensate by Lighten (sometimes still impossible to achieve the desired result)

@ilia3101: What is your proposition? Whole new approach or refined current solution?

**Comment by masc4ii** [comment]
I hope the 2nd. I would miss the other sliders :-D
Btw: GUI is prepared for Shadows & Highlights.

**Comment by bouncyball-git** [comment]
Is there any news?

**Comment by masc4ii** [comment]
https://github.com/ilia3101/MLV-App/commit/fb98493b4269eae10968936e2840af5077da1c04
See comments.

**Comment by ilia3101** [comment]
Ah just noticed this. Current sliders will stay, I like the control they give. I think the reason for not being able to achieve desired result sometimes (bouncyball mentioned) is the lack of shadows/highlights adjustments - it will be perfect when they work (I'm still getting an extremely weird crash every time)

**Comment by bouncyball-git** [comment]
@ilia3101 

Can you commit the changes for us to play with them? Even if it crashes. Or at least make new branch :)

**Comment by ilia3101** [comment]
Ok I'll commit today. In the current state, the crash prevents from even getting a single frame, so I'll try and get that working a bit then commit (highlight reconstruction still broken)

**Comment by bouncyball-git** [comment]
OK thanx :)

**Comment by ilia3101** [comment]
I'm sorry, just gave up in frustration for a while. But I'm back to this now, trying to figure out what is causing the crash. It just makes no sense and I can't even get any images from it right now.

**Comment by masc4ii** [comment]
We still have #74 , but after that we could make a release.

**Comment by masc4ii** [comment]
@ilia3101 : one question: is the Lighten slider more or less the same as the Shadows slider will be? For me - if I compare ACR, Resolve and MLV App there is not much difference...

**Comment by ilia3101** [comment]
```
is the Lighten slider more or less the same as the Shadows slider will be? 
```
No, it is a global adjustment and reduces contrast, the shadows/highlights adjustments will use something similar a blur mask and retain contrast. I have had to restart this as first attempt went badly wrong. Will commit some kinda working changes soon.

Highlight reconstruction may also get changed up completely.

**Comment by masc4ii** [comment]
@ilia3101 : Shadows & Highlights works in QtGui. What is your expected range? I set it -1.5..1.5. And I have an initialization problem: I have to set one of the two values to see a picture - if not, all is black. Do you have the same problem?

**Comment by masc4ii** [comment]
If I comment out this, it works fine (plus { and } ):
```
if (!( processing->shadows_highlights.highlights <  0.01
        && processing->shadows_highlights.highlights > -0.01
        && processing->shadows_highlights.shadows    <  0.01
        && processing->shadows_highlights.shadows    > -0.01))
```

**Comment by bouncyball-git** [comment]
Now there is some more room for tweaking. Highlights slider is great :)

**Comment by masc4ii** [comment]
Both sliders are great! Thank you @ilia3101 !

**Comment by bouncyball-git** [comment]
Great job! @ilia3101

**Comment by bouncyball-git** [comment]
@masc: What about black images? (also black or half black or corrupted thumbnails?)

**Comment by masc4ii** [comment]
Read above, there is a temp solution. But I think Ilia should fix, because he knows what is right and what not...

**Comment by bouncyball-git** [comment]
What I meant above that highlight slider is working on exact higlight range, but shadows slider in addition affects a whole image a lot, not only shadow range.

@ilia3101: can this be narrowed to more shadowish range somehow?

**Comment by masc4ii** [comment]
@ilia3101 : do you have an idea how to get highlights + highlight reconstruction together to work? I was watching the code and I think I understood what the problem is... but do you have a plan how to fix that?

**Comment by bouncyball-git** [comment]
Hey guys did you noticed that if highlights or shadows slider touched/changed playback speed drops by half?

**Comment by masc4ii** [comment]
Yes, this is because if shadows/highlights != 0, there has to be calculated a blurred picture - this needs time and is single threaded atm.

**Comment by bouncyball-git** [comment]
I guessed that it is not multi threaded yet. But why we need to blur the image for highlight/shadow ranges to change? I'm lost :)

**Comment by masc4ii** [comment]
Set highlights to -100 and shadows to +100. The pic is ugly, but you'll see the unsharp mask.

**Comment by masc4ii** [comment]
In raw_processing.c, change the out commenting in line 228 and 229... then it is fast... but looks terrible - so with unsharp mask it looks better.

**Comment by masc4ii** [comment]
Hej @ilia3101 ,
could this make the blur faster? We would have to make it from C++ to C... : 
http://vitiy.info/stackblur-algorithm-multi-threaded-blur-for-cpp/

**Comment by masc4ii** [comment]
I thought I can take the multithreaded standard C stackblur.c/.h from here, rewrite some lines (because it uses X11 libs, I don't want to use them)... but unfortunately I get only crashes - or no blurred picture 😄 
https://github.com/aario/slock-blur
Do you like to try?! Maybe I am to stupid for that 😃

**Comment by ilia3101** [comment]
```
Set highlights to -100 and shadows to +100. The pic is ugly, but you'll see the unsharp mask.
```

Well Qt app seems to be overdoing the value by 2x... (both were set to -100 and +100)
![screen shot 2018-03-24 at 20 15 52](https://user-images.githubusercontent.com/23642861/37868547-4e5649e4-2fa0-11e8-8e9a-3e0611c2f88e.png)

**Comment by masc4ii** [comment]
Your right @ilia3101 , I overdo it by 150% - but I liked that you can do a kind of HDR look with that. I think I won't do this often, but maybe there is someone out there who needs that...
`processingSetShadows( m_pProcessingObject, position * 1.5 / 100.0 );`
`processingSetHighlights( m_pProcessingObject, position * 1.5 / 100.0 );`

**Comment by ilia3101** [comment]
It's a choice: either give user the freedom to make it look horrible, or restrict them so they can't over do it. Difficult question for me.

I guess this is alright as it _could_ be useful occasionally.

That blur algorithm would be interesting to test out, and see if it's faster than my already world's simplest blur. Looks kinda fiddly to implement though. If we don't successfully implement this I can always make mine be multi threaded.

**Comment by masc4ii** [comment]
Yes... often you can make look it horrible. But especially in night shots I got sometimes nice results with highlights at -100 (*1.5). For shadows... hmmm... mostly I stay in the range 0..20. 

Maybe that code helps you. For me it was crashing in the horizontal function. 
But if you have a better idea how to make your code multi threaded - maybe you are faster with that... here you are already at 3x16bit.

**Comment by masc4ii** [comment]
The easiest way to make your function faster could be to calc each color in parallel. Then it would be around 3x faster - and there is nearly no overhead. Or what do you think?!

**Comment by masc4ii** [comment]
@ilia3101 : I found another multithreaded blur, this time in standard C! And this one seems to work without any libraries (the BMP library can be deleted and replaced with our makros and variables).
https://github.com/ha107642/Parallel-Blur/blob/master/blurpt.c
from here:
https://github.com/ha107642/Parallel-Blur
If I find some time, I would give it a try - or do you like to try it out?

**Comment by masc4ii** [comment]
@ilia3101 : I converted the blur code for our project. It compiles and does not crash for me. It runs very fast too. But I can't see any blur... do you like to try out? Add both files in processing folder, include header and exchange `blur_image(..)` with `blurImageThreaded(..)`.
```
#include <stdlib.h>
#include <pthread.h>
#include "blurThreaded.h"

uint16_t m_width, m_height;
uint8_t m_radius;
uint16_t *m_input;
uint16_t *m_output;
uint8_t m_numThreads;

void* horizontal_blur(void* tnum) 
{
    uint16_t width, height;
    uint16_t r, g, b;
    uint16_t* pixel;

    int thread_num = (int)tnum;
    int radius = m_radius;
    int blur_width = 2*radius+1;
    int i, x, y;
    int next, prev;
    int start, end;
    float blur_width_divisor = 1.0 / blur_width;

    //Get dimensions
    width = m_width;
    height = m_height;

    uint16_t bytes_per_row;
    uint16_t bytes_per_pixel;
    uint16_t* data;
    uint16_t* outdata;

    bytes_per_pixel = 3;// * sizeof( uint16_t );
    bytes_per_row = width * 3;// * sizeof( uint16_t );
    data = m_input;
    outdata = m_output;

    start = (height / m_numThreads) * thread_num;
    if (thread_num == m_numThreads - 1)
        end = height;
    else
        end = (height / m_numThreads)*(thread_num + 1);

    for (y = start; y < end; ++y) 
    {
		pixel = data + ((height - y - 1) * bytes_per_row);
		r = (*( pixel + 2 ))*radius;
		g = (*( pixel + 1 ))*radius;
		b = (*( pixel + 0 ))*radius;

        for (i = 0; i < radius; ++i)
        {
        	//Calculate the sum of all pixels' red, green and blue values
            //for the blur width.
            pixel = data + ((height - y - 1) * bytes_per_row + ((i<0) ? 0 : i) * bytes_per_pixel);
			r += *( pixel + 2 );
			g += *( pixel + 1 );
			b += *( pixel + 0 );
        }

        for (x = 0; x < width; ++x) 
        {
        	// //Set the current pixels value to the sum of calulated pixels divided by blur width.
			pixel = outdata + ((height - y - 1) * bytes_per_row + x * bytes_per_pixel);			
			/* Note: colors are stored in BGR order */
			*( pixel + 2 ) = r * blur_width_divisor;
			*( pixel + 1 ) = g * blur_width_divisor;
			*( pixel + 0 ) = b * blur_width_divisor;

         	//Check if edges of blur width are out of bounds.
           	next = (x+radius+1 > width - 1) ? width - 1 : x+radius+1; //If at the end of array (out of bounds)
         	prev = (x-radius < 0) ?  0 : x-radius; //Again, check if out of bounds.

            //Add the next pixel to the pixel sums.
		 	pixel = data + ((height - y - 1) * bytes_per_row + next * bytes_per_pixel);
			r += *( pixel + 2 );
			g += *( pixel + 1 );
			b += *( pixel + 0 );

            //... And remove the previous one.
			pixel = data + ((height - y - 1) * bytes_per_row + prev * bytes_per_pixel);
			r -= *( pixel + 2 );
			g -= *( pixel + 1 );
			b -= *( pixel + 0 );
		}
    }
    return NULL;
}

//This function is MUCH faster than the simple way, since this function
//will not cause nearly as many cache misses. 
void* vertical_blur(void* tnum) 
{
    uint16_t width, height;
    uint16_t *r, *g, *b;
    uint16_t* pixel;

    int thread_num = (int)tnum;
    int radius = m_radius;
    int blur_width = 2*radius+1;
    int i, x, y;
    int next, prev;
    int start, end;
    float blur_width_divisor = 1.0 / blur_width;

    //Get dimensions
    width = m_width;
    height = m_height;

    uint16_t bytes_per_row;
    uint16_t bytes_per_pixel;
    uint16_t* data;
    uint16_t* outdata;

    bytes_per_pixel = 3;// * sizeof( uint16_t );
    bytes_per_row = width * 3;// * sizeof( uint16_t );
    data = m_input;
    outdata = m_output;

    r = malloc(width*sizeof(uint16_t*));
    g = malloc(width*sizeof(uint16_t*));
    b = malloc(width*sizeof(uint16_t*));

    start = (width / m_numThreads) * thread_num;
    if (thread_num == m_numThreads - 1)
        end = width;
    else
        end = (width / m_numThreads)*(thread_num + 1);

    for (x = start; x < end; ++x) 
    {
    	pixel = data + ((height - 1) * bytes_per_row + x * bytes_per_pixel);
		r[x] = (*( pixel + 2 ))*radius;
		g[x] = (*( pixel + 1 ))*radius;
		b[x] = (*( pixel + 0 ))*radius;
    }

    for (i = 0; i < radius; ++i)
    {
    	for (x = start; x < end; ++x) 
    	{
			//Calculate the sum of all pixels' red, green and blue values
			//for the blur width.	
	    	pixel = data + ((height - i - 1) * bytes_per_row + x * bytes_per_pixel);
			r[x] += *( pixel + 2 );
			g[x] += *( pixel + 1 );
			b[x] += *( pixel + 0 );
    	}
    }

    for (y = 0; y < height; ++y) 
    {    
    	//Check if edges of blur width are out of bounds.
        next = (y+radius+1 > height - 1) ? height - 1 : y+radius+1; //If at the end of array (out of bounds)
        prev = (y-radius < 0) ?  0 : y-radius; //Again, check if out of bounds.

    	for (x = start; x < end; ++x) 
    	{
	       	//Set the current pixels value to the sum of calulated pixels divided by blur width.
			pixel = outdata + ((height - y - 1) * bytes_per_row + x * bytes_per_pixel);

			/* Note: colors are stored in BGR order */
			*( pixel + 2 ) = r[x] * blur_width_divisor;
			*( pixel + 1 ) = g[x] * blur_width_divisor;
			*( pixel + 0 ) = b[x] * blur_width_divisor;

	        //Add the next pixel to the pixel sums.
			pixel = data + ((height - next - 1) * bytes_per_row + x * bytes_per_pixel);
			r[x] += *( pixel + 2 );
			g[x] += *( pixel + 1 );
			b[x] += *( pixel + 0 );

	        //... And remove the previous one.
			pixel = data + ((height - prev - 1) * bytes_per_row + x * bytes_per_pixel);
			r[x] -= *( pixel + 2 );
			g[x] -= *( pixel + 1 );
			b[x] -= *( pixel + 0 );
		}
    }
    free(r);
    free(g);
    free(b);
    return NULL;
}

void blurImageThreaded( uint16_t* output, uint16_t* input, uint16_t imageX, uint16_t imageY, uint8_t radius, uint8_t numThreads )
{
    int i;

    m_numThreads = numThreads;
    pthread_t thread[numThreads];
    pthread_attr_t attr;

    m_input = input;
    m_output = output;

    m_width = imageX;
    m_height = imageY;
    m_radius = radius;

    pthread_attr_init(&attr);
    pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE);

    /* Create and run the threads */
    for (i=0; i<numThreads; i++)
            pthread_create(&thread[i], &attr, horizontal_blur, (void *) i);

    /* Join the threads - barrier*/
    for (i=0; i<numThreads; i++)
            pthread_join(thread[i], NULL);

    for (i=0; i<numThreads; i++)
            pthread_create(&thread[i], &attr, vertical_blur, (void *) i);

    for (i=0; i<numThreads; i++)
            pthread_join(thread[i], NULL);

    return;
}
```

```
#ifndef BLURTHREADED_H
#define BLURTHREADED_H

void blurImageThreaded(uint16_t* output, uint16_t* input, uint16_t imageX, uint16_t imageY, uint8_t radius, uint8_t numThreads );

#endif // BLURTHREADED_H
```

**Comment by masc4ii** [comment]
What a shame... tried it on the output to see what it does... something seems to be really wrong 😢

**Comment by bouncyball-git** [comment]
You mean it does nothing? :)

**Comment by masc4ii** [comment]
I tried it out in `getMlvProcessedFrame16()` to better see what it does on output. Histogram:
![bildschirmfoto 2018-04-14 um 10 51 10](https://user-images.githubusercontent.com/30245296/38766399-fd8a807e-3fd1-11e8-8fc5-503535fa5667.png)
Then I exported the png and photoshopped it to see something:
![m23-1205_frame_1](https://user-images.githubusercontent.com/30245296/38766414-381026d6-3fd2-11e8-8645-e407df157021.png)
You see the boot in the scene? 😄

**Comment by masc4ii** [comment]
Maybe I should try the same thread principle with the blur function by @ilia3101. Here we know it works and it is also way shorter and better understandable...

**Comment by masc4ii** [comment]
Please try it out, friends! I now combined the threading of the code above with the blur code from @ilia3101 . On my dualcode it slightly uses more CPU and is not way faster... but I have a blurred correct image. Maybe on you machines it is faster! ;) Check out latest commit!

**Comment by bouncyball-git** [comment]
Surrealistic boat! I love it :)
Will check changes when at home.

**Comment by masc4ii** [comment]
I tried it on Windows too: the more Cores you have, the faster it is - as expected. On i7 I now loose 1-2fps when using Highlights/Shadows! If you have no more problems, I think we can close here!
